n_helpers.c

```diff
@@ -74,9 +79,204 @@ static short normalYearDaysByMonth[] = {0, 31, 59, 90, 120, 151, 181, 212, 243,
 static const char *daynames[] = {"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"};
 
 // Forwards
+NOTE_C_STATIC void setTime(JTIME seconds);
 NOTE_C_STATIC bool timerExpiredSecs(uint32_t *timer, uint32_t periodSecs);
 NOTE_C_STATIC int ytodays(int year);
 
+//**************************************************************************/
+/*!
+  @brief  Receive a large binary object from the Notecard's binary buffer
+  @param  buffer A buffer to hold the result
+  @param  bufLen The total length of the buffer
+  @returns  NULL on success, else an error string pointer.
+  @note  Buffers are decoded in place, the original contents of the buffer will
+         be modified.
+*/
+/**************************************************************************/
+const char * NoteBinaryReceive(uint8_t * buffer, size_t bufLen)
+{
+    // Issue a "card.binary" request.
+    J *rsp = NoteRequestResponse(NoteNewRequest("card.binary"));
+    if (!rsp) {
+        return ERRSTR("unable to issue binary request", c_err);
+    }
+
+    // Ensure the transaction doesn't return an error
+    // to confirm the binary feature is available
+    if (NoteResponseError(rsp)) {
+        JDelete(rsp);
+        return ERRSTR("feature not implemented", c_bad);
+    }
+
+    // Examine "cobs" from the response to evaluate the space
+    // required to hold the COBS encoded data on the Notecard.
+    const size_t cobs = JGetInt(rsp,"cobs");
+    JDelete(rsp);
+    if (!cobs) {
+        return ERRSTR("no data on notecard", c_err);
+    }
+    if (cobs > bufLen) {
```

Comment:
I know we already discussed this, but it still rubs me the wrong way that the caller would have to either

1. Allocate a buffer that they assume will be large enough.
2. Do `card.binary` manually themselves, get the length of buffer needed, allocate that, and finally call this function.

One solution I've seen in other open-source C libraries:

Have this function take a  `uint8_t **` instead of a `uint8_t *`. `buffer == NULL` is an error. `buffer != NULL && *buffer == NULL` means allocate the buffer for me and return it in `*buffer`. `buffer != NULL && *buffer != NULL` means I, the caller, have allocated the buffer already (heap, stack -- doesn't matter), so use that. `size_t bufLen` becomes `size_t * bufLen`. In the case this function allocates the buffer for the caller, we return the length in `*bufLen`. In the case the caller is providing the buffer, we read the length of said buffer from `*bufLen`.

If that seems too complicated, that's fair. There's a middle ground somewhere, I'm sure. But I do think there's some unnecessary friction here.

Reply:
I want to run this by Ray, but I think we need a separate API `NoteBinaryAllocRecieve()`, so it's explicit who's responsible for cleaning up the buffer.

Reply:
before I do that, I do want to think through how this may facilitate with Mat's feedback down below. Maybe leaving the door open for a partial reception.